File: studio/AuraStudio/Sources/AuraStudio/Models/CatalogSummary.swift
import Foundation

/// Fase 24 (PLAN-UX.md): contadores y bytes por tipo tras un sync, para
/// que "Acerca de" en el firmware pueda mostrar cuanto hay realmente en
/// el dispositivo. El firmware no tiene parser de JSON (su unico formato
/// de config es el `key: value` plano que ya usa `aura.cfg`, ver
/// `aura_settings_load`/`settings_parseline`) -- en vez de escribir un
/// parser JSON en C para un solo archivo, `CatalogSummaryWriter` emite
/// ese mismo formato plano. `sync_manifest.json` (SyncManifest, en este
/// mismo directorio) sigue siendo JSON porque es enteramente interno de
/// Studio -- el firmware nunca lo lee.
struct CatalogTypeSummary: Equatable {
    var count: Int = 0
    var bytes: Int64 = 0
}

struct CatalogSummary: Equatable {
    var music = CatalogTypeSummary()
    var video = CatalogTypeSummary()
    var photo = CatalogTypeSummary()
    var playlistCount = 0

    /// D-283 (PLAN-about-fixes.md E2/Q6): conteos por categoria dentro de
    /// video/foto -- Studio ya clasifica cada item al importar
    /// (MediaCategory para video, MediaCategoryHeuristics.classifyPhoto
    /// para foto); el firmware no tiene base de datos de video ni parser
    /// EXIF, asi que no puede clasificar nada por si solo (Rockbox no
    /// tiene esa fuente). Este es el mismo canal `sync_summary.cfg` que
    /// ya existia para bytes/conteos totales, solo con mas lineas -- el
    /// firmware SOLO lee, nunca re-clasifica. "videoClips" = la
    /// categoria `.videos` de MediaCategory (video sin clasificar a
    /// mano por el usuario, que el Estado 2 del firmware llama
    /// "videoclips" siguiendo el encargo original del dueño).
    var videoMovies = 0
    var videoSeries = 0
    var videoClips = 0
    var photoImages = 0
    var photoPhotos = 0
    var photoAI = 0
}

/// Lee de vuelta el mismo archivo que escribe `CatalogSummaryWriter`.
/// Studio lo usa para contar lo que YA hay en el iPod sin recorrer el
/// disco entero: el resumen lo dejo el ultimo sync, y el firmware lo lee
/// igual para su pantalla "Acerca de".
enum CatalogSummaryReader {
    static func parse(_ text: String) -> CatalogSummary {
        var values: [String: Int64] = [:]
        for line in text.split(separator: "\n") {
            let parts = line.split(separator: ":", maxSplits: 1)
            guard parts.count == 2 else { continue }
            let key = parts[0].trimmingCharacters(in: .whitespaces)
            let raw = parts[1].trimmingCharacters(in: .whitespaces)
            guard let value = Int64(raw) else { continue }
            values[key] = value
        }

        var summary = CatalogSummary()
        summary.music = CatalogTypeSummary(count: Int(values["music_count"] ?? 0),
                                            bytes: values["music_bytes"] ?? 0)
        summary.video = CatalogTypeSummary(count: Int(values["video_count"] ?? 0),
                                            bytes: values["video_bytes"] ?? 0)
        summary.photo = CatalogTypeSummary(count: Int(values["photo_count"] ?? 0),
                                            bytes: values["photo_bytes"] ?? 0)
        summary.playlistCount = Int(values["playlist_count"] ?? 0)
        summary.videoMovies = Int(values["video_movies_count"] ?? 0)
        summary.videoSeries = Int(values["video_series_count"] ?? 0)
        summary.videoClips = Int(values["video_clips_count"] ?? 0)
        summary.photoImages = Int(values["photo_images_count"] ?? 0)
        summary.photoPhotos = Int(values["photo_photos_count"] ?? 0)
        summary.photoAI = Int(values["photo_ai_count"] ?? 0)
        return summary
    }
}

enum CatalogSummaryWriter {
    static func serialize(_ summary: CatalogSummary) -> String {
        """
        music_count: \(summary.music.count)
        music_bytes: \(summary.music.bytes)
        video_count: \(summary.video.count)
        video_bytes: \(summary.video.bytes)
        photo_count: \(summary.photo.count)
        photo_bytes: \(summary.photo.bytes)
        playlist_count: \(summary.playlistCount)
        video_movies_count: \(summary.videoMovies)
        video_series_count: \(summary.videoSeries)
        video_clips_count: \(summary.videoClips)
        photo_images_count: \(summary.photoImages)
        photo_photos_count: \(summary.photoPhotos)
        photo_ai_count: \(summary.photoAI)

        """
    }
}

File: studio/windows/AuraStudio.Core/CatalogSummary.cs
namespace AuraStudio.Core;

/// <summary>
/// Contadores y bytes por tipo de contenido tras un sync, para que
/// "Acerca de" en el firmware pueda mostrar cuánto hay realmente en
/// el dispositivo. El firmware no tiene parser de JSON (su único formato
/// de config es el `key: value` plano que ya usa `aura.cfg`), así que
/// `CatalogSummaryWriter` emite ese mismo formato plano.
/// </summary>
public struct CatalogTypeSummary : IEquatable<CatalogTypeSummary>
{
    public int Count;
    public long Bytes;

    public bool Equals(CatalogTypeSummary other) =>
        Count == other.Count && Bytes == other.Bytes;

    public override bool Equals(object? obj) =>
        obj is CatalogTypeSummary other && Equals(other);

    public override int GetHashCode() => HashCode.Combine(Count, Bytes);

    public static bool operator ==(CatalogTypeSummary left, CatalogTypeSummary right) => left.Equals(right);
    public static bool operator !=(CatalogTypeSummary left, CatalogTypeSummary right) => !left.Equals(right);
}

/// <summary>
/// Resumen completo del catálogo en el iPod: conteos y bytes por tipo,
/// más conteos por subcategoría de video/foto que Studio clasificó al
/// importar pero el firmware no puede clasificar por sí solo.
/// </summary>
public struct CatalogSummary : IEquatable<CatalogSummary>
{
    public CatalogTypeSummary Music;
    public CatalogTypeSummary Video;
    public CatalogTypeSummary Photo;
    public int PlaylistCount;

    /// <summary>video_movies, video_series, video_clips: subcategorías de video clasificadas por Studio.</summary>
    public int VideoMovies;
    public int VideoSeries;
    public int VideoClips;

    /// <summary>photo_images, photo_photos, photo_ai: subcategorías de foto clasificadas por Studio.</summary>
    public int PhotoImages;
    public int PhotoPhotos;
    public int PhotoAI;

    public bool Equals(CatalogSummary other) =>
        Music.Equals(other.Music) &&
        Video.Equals(other.Video) &&
        Photo.Equals(other.Photo) &&
        PlaylistCount == other.PlaylistCount &&
        VideoMovies == other.VideoMovies &&
        VideoSeries == other.VideoSeries &&
        VideoClips == other.VideoClips &&
        PhotoImages == other.PhotoImages &&
        PhotoPhotos == other.PhotoPhotos &&
        PhotoAI == other.PhotoAI;

    public override bool Equals(object? obj) =>
        obj is CatalogSummary other && Equals(other);

    public override int GetHashCode() =>
        HashCode.Combine(HashCode.Combine(Music, Video, Photo, PlaylistCount,
            VideoMovies, VideoSeries, VideoClips),
            PhotoImages, PhotoPhotos, PhotoAI);

    public static bool operator ==(CatalogSummary left, CatalogSummary right) => left.Equals(right);
    public static bool operator !=(CatalogSummary left, CatalogSummary right) => !left.Equals(right);
}

/// <summary>
/// Lee de vuelta el mismo archivo que escribe <see cref="CatalogSummaryWriter"/>.
/// Studio lo usa para contar lo que YA hay en el iPod sin recorrer el
/// disco entero: el resumen lo dejó el último sync, y el firmware lo lee
/// igual para su pantalla "Acerca de".
/// </summary>
public static class CatalogSummaryReader
{
    public static CatalogSummary Parse(string text)
    {
        var values = new Dictionary<string, long>();

        foreach (var line in text.Split('\n'))
        {
            var parts = line.Split(':', 2);
            if (parts.Length != 2) continue;

            var key = parts[0].Trim();
            var raw = parts[1].Trim();

            if (!long.TryParse(raw, out var value)) continue;

            values[key] = value;
        }

        return new CatalogSummary
        {
            Music = new CatalogTypeSummary
            {
                Count = (int)(values.TryGetValue("music_count", out var mc) ? mc : 0),
                Bytes = values.TryGetValue("music_bytes", out var mb) ? mb : 0
            },
            Video = new CatalogTypeSummary
            {
                Count = (int)(values.TryGetValue("video_count", out var vc) ? vc : 0),
                Bytes = values.TryGetValue("video_bytes", out var vb) ? vb : 0
            },
            Photo = new CatalogTypeSummary
            {
                Count = (int)(values.TryGetValue("photo_count", out var pc) ? pc : 0),
                Bytes = values.TryGetValue("photo_bytes", out var pb) ? pb : 0
            },
            PlaylistCount = (int)(values.TryGetValue("playlist_count", out var pl) ? pl : 0),
            VideoMovies = (int)(values.TryGetValue("video_movies_count", out var vm) ? vm : 0),
            VideoSeries = (int)(values.TryGetValue("video_series_count", out var vs) ? vs : 0),
            VideoClips = (int)(values.TryGetValue("video_clips_count", out var vcl) ? vcl : 0),
            PhotoImages = (int)(values.TryGetValue("photo_images_count", out var pi) ? pi : 0),
            PhotoPhotos = (int)(values.TryGetValue("photo_photos_count", out var pp) ? pp : 0),
            PhotoAI = (int)(values.TryGetValue("photo_ai_count", out var pa) ? pa : 0)
        };
    }
}

/// <summary>
/// Escribe el formato plano `key: value` que el firmware puede leer
/// sin necesidad de un parser JSON.
/// </summary>
public static class CatalogSummaryWriter
{
    public static string Serialize(CatalogSummary summary) =>
        $"music_count: {summary.Music.Count}\n" +
        $"music_bytes: {summary.Music.Bytes}\n" +
        $"video_count: {summary.Video.Count}\n" +
        $"video_bytes: {summary.Video.Bytes}\n" +
        $"photo_count: {summary.Photo.Count}\n" +
        $"photo_bytes: {summary.Photo.Bytes}\n" +
        $"playlist_count: {summary.PlaylistCount}\n" +
        $"video_movies_count: {summary.VideoMovies}\n" +
        $"video_series_count: {summary.VideoSeries}\n" +
        $"video_clips_count: {summary.VideoClips}\n" +
        $"photo_images_count: {summary.PhotoImages}\n" +
        $"photo_photos_count: {summary.PhotoPhotos}\n" +
        $"photo_ai_count: {summary.PhotoAI}\n";
}
